src/Database/LSMTree/Internal/IndexCompact.hs

```diff
@@ -385,8 +386,22 @@ newtype PageNo = PageNo { unPageNo :: Int }
   deriving stock (Show, Eq, Ord)
   deriving newtype NFData
 
--- TODO: Turn into newtype
-type NumPages = Int
+-- | The number of pages contained by an index or other paging data-structure.
+--
+-- Note: This is a 0-based number; take care to ensure arithmetic underflow
+-- does not occur during subtraction operations!
+newtype NumPages = NumPages Word
+  deriving stock (Eq, Ord, Show)
+  deriving newtype (NFData)
+
+-- | A type-safe "unwrapper" for 'NumPages'. Use this accessor whenever you want
+-- to convert 'NumPages' to a more versitile number type.
+{-# INLINE getNumPages #-}
+{-# SPECIALIZE getNumPages :: NumPages -> Int #-}
+{-# SPECIALIZE getNumPages :: NumPages -> Word #-}
+{-# SPECIALIZE getNumPages :: NumPages -> Word64 #-}
```

Comment:
Just the `INLINE` should be sufficient for specialisation to happen at the use site (in most cases)

Reply:
Also, AFAIK the `INLINE` pragma doesn't apply to the specialised versions of the function

Reply:
I wasn't sure so I leaned towards the side of caution. Is there a downside to leaving the specializations (larger object file)?

Reply:
I removed the `SPECIALIZE` statements.